File: aoc2015/clojure/src/aoc2015/day06.rb

def get_starting_grid
  start_grid = {} 
  1000.times do |x|
    1000.times do |y|
      start_grid["#{x}-#{y}"] = false
    end
  end
  start_grid
end

def parse(s) 
  /(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)/.match(s)
end

def change(x1, y1, x2, y2, state, grid)
  (x1..x2).each do |x|
    (y1..y2).each do |y|
      key = "#{x}-#{y}"
      if state == "turn on" then
        grid[key] = true
      elsif state == "turn off" then
        grid[key] = false
      else
        grid[key] = !grid[key]
      end
    end
  end
end

grid = get_starting_grid()
File.readlines('resources/day06.input').each do |line|
  puts "Reading line \"#{line}\""
  vars = parse(line)
  change(vars[2], vars[3], vars[4], vars[5], vars[1], grid)

end

puts grid.values.count{|i| i}

File: aoc2015/clojure/src/aoc2015/day06.clj
(ns aoc2015.day06)

(def x-axis (range 1000))
(def y-axis (range 1000))

(def starting-grid
  (concat 
   (first 
    (for [x x-axis]
      (for [y y-axis]
        [(list x y) false])))))

(defn map-change [change state inmap]
  (map (fn [i] 
         (cond 
           (and (= change (first i)) (= state "turn on"))
           (list change true)

           (and (= change (first i)) (= state "turn off"))
           (list change false)

           (and (= change (first i)) (= state "toggle"))
           (list change (not (second i)))

           :else
           i)) inmap))

(defn make-change [x1 y1 x2 y2 state input]
  (let [change-list (concat 
                     (first 
                      (for [x (range x1 (inc x2))]
                        (for [y (range y1 (inc y2))]
                          (list x y)))))]
    (loop [[change & t] change-list
           grid input]
      (if (nil? t)
        (map-change change state grid)
        (recur t (map-change change state grid))))))

(defn parse [s] 
  (let [result (re-find #"(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)" s)]
    [(Integer. (nth result 2))
     (Integer. (nth result 3))
     (Integer. (nth result 4))
     (Integer. (nth result 5))
     (second result)]))

(defn process [input s]
  (let [[x1 y1 x2 y2 state] (parse s)]
    (make-change x1 y1 x2 y2 state input)))
